# Importing libraries and Dataset

In [112]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style('darkgrid')
from datetime import datetime
from collections import Counter

In [113]:
df = pd.read_csv('data/Police_Department_Incident_Reports__Historical_2003_to_May_2018.csv')
print(df.shape)
df.head()

(2215024, 13)


IncidntNum         Category                                   Descript  \
0   150060275     NON-CRIMINAL                              LOST PROPERTY   
1   150098210          ROBBERY                      ROBBERY, BODILY FORCE   
2   150098210          ASSAULT       AGGRAVATED ASSAULT WITH BODILY FORCE   
3   150098210  SECONDARY CODES                          DOMESTIC VIOLENCE   
4   150098226        VANDALISM  MALICIOUS MISCHIEF, VANDALISM OF VEHICLES   

  DayOfWeek        Date   Time  PdDistrict Resolution  \
0    Monday  01/19/2015  14:00     MISSION       NONE   
1    Sunday  02/01/2015  15:45  TENDERLOIN       NONE   
2    Sunday  02/01/2015  15:45  TENDERLOIN       NONE   
3    Sunday  02/01/2015  15:45  TENDERLOIN       NONE   
4   Tuesday  01/27/2015  19:00    NORTHERN       NONE   

                       Address           X          Y  \
0        18TH ST / VALENCIA ST -122.421582  37.761701   
1  300 Block of LEAVENWORTH ST -122.414406  37.784191   
2  300 Block of LEAVENWORTH ST -122.414406  37.784191   
3  300 Block of LEAVENWORTH ST -122.414406  37.784191   
4       LOMBARD ST / LAGUNA ST -122.431119  37.800469   

                                Location            PdId  
0   (37.7617007179518, -122.42158168137)  15006027571000  
1  (37.7841907151119, -122.414406029855)  15009821003074  
2  (37.7841907151119, -122.414406029855)  15009821004014  
3  (37.7841907151119, -122.414406029855)  15009821015200  
4  (37.8004687042875, -122.431118543788)  15009822628160

## 1. Inspect data for any duplicates

To-dos:
* To calculate total number of crimes, make sure not to double count
* Double counting is alright if calculating based on crime types but delete if both incident_num and category are the same

From the metadata, we understand that `PdId` refers to the unique identifier for use in update and insert operations for the dataset, hence we can drop the column.

In [114]:
df=df.drop(['PdId'], axis=1)

In [115]:
df.head()

IncidntNum         Category                                   Descript  \
0   150060275     NON-CRIMINAL                              LOST PROPERTY   
1   150098210          ROBBERY                      ROBBERY, BODILY FORCE   
2   150098210          ASSAULT       AGGRAVATED ASSAULT WITH BODILY FORCE   
3   150098210  SECONDARY CODES                          DOMESTIC VIOLENCE   
4   150098226        VANDALISM  MALICIOUS MISCHIEF, VANDALISM OF VEHICLES   

  DayOfWeek        Date   Time  PdDistrict Resolution  \
0    Monday  01/19/2015  14:00     MISSION       NONE   
1    Sunday  02/01/2015  15:45  TENDERLOIN       NONE   
2    Sunday  02/01/2015  15:45  TENDERLOIN       NONE   
3    Sunday  02/01/2015  15:45  TENDERLOIN       NONE   
4   Tuesday  01/27/2015  19:00    NORTHERN       NONE   

                       Address           X          Y  \
0        18TH ST / VALENCIA ST -122.421582  37.761701   
1  300 Block of LEAVENWORTH ST -122.414406  37.784191   
2  300 Block of LEAVENWORTH ST -122.414406  37.784191   
3  300 Block of LEAVENWORTH ST -122.414406  37.784191   
4       LOMBARD ST / LAGUNA ST -122.431119  37.800469   

                                Location  
0   (37.7617007179518, -122.42158168137)  
1  (37.7841907151119, -122.414406029855)  
2  (37.7841907151119, -122.414406029855)  
3  (37.7841907151119, -122.414406029855)  
4  (37.8004687042875, -122.431118543788)

In [116]:
# check unique no. of rows (corresponding to no. of incident numbers)
print("No. of rows:",df.shape[0])

No. of rows: 2215024


In [117]:
# check unique no. of incidents
print("No. of unique incidents:", df['IncidntNum'].nunique())

No. of unique incidents: 1746914


In [118]:
df.shape[0]-df['IncidntNum'].nunique()

468110

Hence we can conclude that there were 468110 duplicates in incident_num. Let's take a look at an example of a duplicate incident. The logic is that, if the duplicates arise due to the incident being categorized into various incidents of crime, we can retain the double-counting; however, if the duplicates are merely a result of same incident_num and category, we should delete these rows.

In [119]:
incidentNumList=Counter(df['IncidntNum']) # obtain dictionary of all unique key-value pairs 

In [120]:
incidentNumList

Counter({150060275: 1,
         150098210: 3,
         150098226: 1,
         150098232: 1,
         150098248: 2,
         150098254: 1,
         150098260: 1,
         150098345: 4,
         150098367: 1,
         150098373: 2,
         150098389: 1,
         150098395: 1,
         150098408: 1,
         150098414: 4,
         160919032: 1,
         150098420: 3,
         150098436: 1,
         150098442: 1,
         150098458: 3,
         140009459: 2,
         150098464: 1,
         150098470: 1,
         150098486: 1,
         150098492: 2,
         150098505: 1,
         150098511: 1,
         150098527: 3,
         150098533: 1,
         150098549: 1,
         150098555: 1,
         150098561: 1,
         150098577: 1,
         150098583: 1,
         150098599: 1,
         150098602: 3,
         150098618: 1,
         150098624: 1,
         150098630: 1,
         150098652: 2,
         150098668: 1,
         150098674: 3,
         150098680: 1,
         150098696: 1,
         15

In [121]:
allIncidentNumbers=df['IncidntNum'].unique().tolist()

In [122]:
len(incidentNumList) 

1746914

In [123]:
threshold=1
duplicateIncidentList=[]
isolatedIncidentList=[]
for incident, incidentCount in incidentNumList.items():
    if incidentCount>threshold:
        duplicateIncidentList.append(incident)
    elif incidentCount==threshold:
        isolatedIncidentList.append(incident)

# shorter code is this: 
# isolatedIncidents=[incd for incd, incidentCount in incidentNumList.items() if incidentCount == threshold]
# duplicateIncidents=[incd for incd, incidentCount in incidentNumList.items() if incidentCount > threshold]
# sumIsolates=len(isolatedIncidents)
# sumDuplicate=len(duplicateIncidents)

In [124]:
len(duplicateIncidentList)

341334

In [125]:
len(isolatedIncidentList)

1405580

In [126]:
'''check random incident_num with non-1s count to see reason for duplicates
    if incident_num and category are the same, then remove duplicates (later)
'''
df[df['IncidntNum']==150098345]

IncidntNum       Category                               Descript  \
10   150098345  LARCENY/THEFT                PETTY THEFT SHOPLIFTING   
11   150098345  DRUG/NARCOTIC         POSSESSION OF METH-AMPHETAMINE   
12   150098345  DRUG/NARCOTIC  POSSESSION OF NARCOTICS PARAPHERNALIA   
13   150098345       WARRANTS                         WARRANT ARREST   

   DayOfWeek        Date   Time PdDistrict      Resolution  \
10    Sunday  02/01/2015  14:00    MISSION  ARREST, BOOKED   
11    Sunday  02/01/2015  14:00    MISSION  ARREST, BOOKED   
12    Sunday  02/01/2015  14:00    MISSION  ARREST, BOOKED   
13    Sunday  02/01/2015  14:00    MISSION  ARREST, BOOKED   

                      Address           X          Y  \
10  1700 Block of HARRISON ST -122.413354  37.769075   
11  1700 Block of HARRISON ST -122.413354  37.769075   
12  1700 Block of HARRISON ST -122.413354  37.769075   
13  1700 Block of HARRISON ST -122.413354  37.769075   

                                 Location  
10  (37.7690748003847, -122.413354187018)  
11  (37.7690748003847, -122.413354187018)  
12  (37.7690748003847, -122.413354187018)  
13  (37.7690748003847, -122.413354187018)

In [127]:
df[df['IncidntNum']==150098210]

IncidntNum         Category                              Descript  \
1   150098210          ROBBERY                 ROBBERY, BODILY FORCE   
2   150098210          ASSAULT  AGGRAVATED ASSAULT WITH BODILY FORCE   
3   150098210  SECONDARY CODES                     DOMESTIC VIOLENCE   

  DayOfWeek        Date   Time  PdDistrict Resolution  \
1    Sunday  02/01/2015  15:45  TENDERLOIN       NONE   
2    Sunday  02/01/2015  15:45  TENDERLOIN       NONE   
3    Sunday  02/01/2015  15:45  TENDERLOIN       NONE   

                       Address           X          Y  \
1  300 Block of LEAVENWORTH ST -122.414406  37.784191   
2  300 Block of LEAVENWORTH ST -122.414406  37.784191   
3  300 Block of LEAVENWORTH ST -122.414406  37.784191   

                                Location  
1  (37.7841907151119, -122.414406029855)  
2  (37.7841907151119, -122.414406029855)  
3  (37.7841907151119, -122.414406029855)

In [128]:
# check the original dataframe against 'duplicateIncidentList'
# when 'duplicateIncidentList' matches df['IncidntNum'], then check if the category is the same
# if the categories are the same, retain one row and drop the others
# if the categories are different, retain all rows 

df_duplicates=df[df['IncidntNum'].isin(allIncidentNumbers)].sort_values(by=['IncidntNum'])
df_duplicates_cleaned = df_duplicates.drop_duplicates(subset=['Category','IncidntNum'], keep="first")
df_duplicates_cleaned # this is the final dataframe without duplicates


IncidntNum                Category  \
493897         3979                WARRANTS   
1209080       10128                WARRANTS   
1490233       10736                WARRANTS   
1393001       38261                WARRANTS   
377806        52205                WARRANTS   
1333777       61397                WARRANTS   
596484        62389           VEHICLE THEFT   
391902        62953                WARRANTS   
1485730       70796                WARRANTS   
864778        73057                WARRANTS   
429957        77231                WARRANTS   
1560901       81737                WARRANTS   
1089774      100622                WARRANTS   
581124       119877                WARRANTS   
1733674      123436  SEX OFFENSES, FORCIBLE   
580589       123436           DRUG/NARCOTIC   
1783453      131912                WARRANTS   
1498591      152576                WARRANTS   
297540       198398                WARRANTS   
601782       248250                WARRANTS   
1476792      259633                WARRANTS   
831206       267153                WARRANTS   
1275712      315235                WARRANTS   
1261745      324684             DRUNKENNESS   
930634       324684                WARRANTS   
1097757      335471                WARRANTS   
1462090      354283                WARRANTS   
1542229      358718                WARRANTS   
712511       359324                WARRANTS   
1084735      363228                WARRANTS   
...             ...                     ...   
1139524   991148424                WARRANTS   
448194    991152237            NON-CRIMINAL   
443970    991161125                WARRANTS   
1197191   991168905                WARRANTS   
1306902   991185686                WARRANTS   
564838    991212154                WARRANTS   
1654666   991231609                WARRANTS   
1359979   991258556                WARRANTS   
1295610   991276821                WARRANTS   
79256     991305157          MISSING PERSON   
403132    991309046                WARRANTS   
441660    991335671                WARRANTS   
777055    991337467                WARRANTS   
520934    991339827                WARRANTS   
79257     991349672          MISSING PERSON   
809727    991400832                WARRANTS   
1695965   991401476                WARRANTS   
1561053   991404577                WARRANTS   
1699957   991417495                WARRANTS   
1147954   991431586                WARRANTS   
607039    991438504                WARRANTS   
1104042   991458988                WARRANTS   
1767722   991473748                WARRANTS   
749616    991487965                WARRANTS   
1234007   991510691                WARRANTS   
1662616   991510691          OTHER OFFENSES   
1529086   991523575                WARRANTS   
543920    991549731                WARRANTS   
1188388   991564488                WARRANTS   
913476    991582377                WARRANTS   

                                             Descript  DayOfWeek        Date  \
493897                                 WARRANT ARREST   Thursday  12/09/2004   
1209080                                WARRANT ARREST     Sunday  12/18/2005   
1490233                                WARRANT ARREST   Thursday  07/15/2004   
1393001                                WARRANT ARREST   Thursday  04/17/2003   
377806                                 WARRANT ARREST   Thursday  02/06/2003   
1333777                                WARRANT ARREST     Sunday  06/29/2003   
596484                              STOLEN MOTORCYCLE    Tuesday  03/30/2004   
391902                                 WARRANT ARREST     Sunday  08/22/2004   
1485730               ENROUTE TO OUTSIDE JURISDICTION   Thursday  07/13/2006   
864778                                 WARRANT ARREST   Thursday  08/26/2004   
429957                                 WARRANT ARREST   Thursday  06/02/2005   
1560901                                WARRANT ARREST   Thursday  12/02/2004   
1089774                                WARRANT ARREST 

In [129]:
df=df_duplicates_cleaned # final dataframe
print("No. of rows and columns of final dataframe: ",df.shape)

No. of rows and columns of final dataframe:  (2057667, 12)


In [130]:
df[df['IncidntNum']==150098345]

IncidntNum       Category                               Descript  \
13   150098345       WARRANTS                         WARRANT ARREST   
12   150098345  DRUG/NARCOTIC  POSSESSION OF NARCOTICS PARAPHERNALIA   
10   150098345  LARCENY/THEFT                PETTY THEFT SHOPLIFTING   

   DayOfWeek        Date   Time PdDistrict      Resolution  \
13    Sunday  02/01/2015  14:00    MISSION  ARREST, BOOKED   
12    Sunday  02/01/2015  14:00    MISSION  ARREST, BOOKED   
10    Sunday  02/01/2015  14:00    MISSION  ARREST, BOOKED   

                      Address           X          Y  \
13  1700 Block of HARRISON ST -122.413354  37.769075   
12  1700 Block of HARRISON ST -122.413354  37.769075   
10  1700 Block of HARRISON ST -122.413354  37.769075   

                                 Location  
13  (37.7690748003847, -122.413354187018)  
12  (37.7690748003847, -122.413354187018)  
10  (37.7690748003847, -122.413354187018)

## 2. Check the unique category of crimes
To-dos:
* Group similar crimes together

In [131]:
# check set of crimes
print(len(set(df['Category'])))
set(df['Category'])

39


{'ARSON',
 'ASSAULT',
 'BAD CHECKS',
 'BRIBERY',
 'BURGLARY',
 'DISORDERLY CONDUCT',
 'DRIVING UNDER THE INFLUENCE',
 'DRUG/NARCOTIC',
 'DRUNKENNESS',
 'EMBEZZLEMENT',
 'EXTORTION',
 'FAMILY OFFENSES',
 'FORGERY/COUNTERFEITING',
 'FRAUD',
 'GAMBLING',
 'KIDNAPPING',
 'LARCENY/THEFT',
 'LIQUOR LAWS',
 'LOITERING',
 'MISSING PERSON',
 'NON-CRIMINAL',
 'OTHER OFFENSES',
 'PORNOGRAPHY/OBSCENE MAT',
 'PROSTITUTION',
 'RECOVERED VEHICLE',
 'ROBBERY',
 'RUNAWAY',
 'SECONDARY CODES',
 'SEX OFFENSES, FORCIBLE',
 'SEX OFFENSES, NON FORCIBLE',
 'STOLEN PROPERTY',
 'SUICIDE',
 'SUSPICIOUS OCC',
 'TREA',
 'TRESPASS',
 'VANDALISM',
 'VEHICLE THEFT',
 'WARRANTS',
 'WEAPON LAWS'}

## 1. Reduce number of crime categories

Group the criminal activities into the following categories:
1. theft
2. public disturbance
3. drug
4. sex
5. suicide
6. kidnap
7. domestic
8. other
9. non-criminal (can delete)

In [132]:
theft = ['BRIBERY',
        'BURGLARY',
        'EXTORTION',        
        'GAMBLING',
        'LARCENY/THEFT',
        'RECOVERED VEHICLE',
        'ROBBERY',
        'TREA',
        'VEHICLE THEFT',
        'STOLEN PROPERTY',
        'ARSON']

fraud = ['FRAUD',
        'EMBEZZLEMENT',
        'FORGERY/COUNTERFEITING']

public = ['TRESPASS',
        'VANDALISM',
        'WARRANTS',
        'LOITERING',
        'ASSAULT',
        'BAD CHECKS',
        'DISORDERLY CONDUCT',
        'DRIVING UNDER THE INFLUENCE',
        'DRUNKENNESS',
        'SUSPICIOUS OCC',
         'LIQUOR LAWS',
         'WEAPON LAWS']

drug =  ['DRUG/NARCOTIC']

sex = ['PORNOGRAPHY/OBSCENE MAT',
        'PROSTITUTION',
        'SEX OFFENSES, FORCIBLE',
        'SEX OFFENSES, NON FORCIBLE']

suicide = ['SUICIDE']

kidnap = ['KIDNAPPING']

domestic = ['MISSING PERSON',
            'RUNAWAY',
            'FAMILY OFFENSES']

others = ['OTHER OFFENSES',
        'SECONDARY CODES']

non_criminal = ['NON-CRIMINAL']

In [133]:
replace_dict = {
                "theft": theft,
                "fraud": fraud,
                "public": public,
                "drug": drug,
                "sex": sex,
                "suicide": suicide,
                "kidnap":kidnap,
                "domestic": domestic,
                "others": others,
                "non_criminal": non_criminal
               }

In [134]:
new_cat = df['Category']

for key, val in replace_dict.items():
    new_cat = new_cat.replace(val, key)

df["New Category"] = new_cat

/Users/clara/anaconda3/envs/data/lib/python3.6/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [135]:
set(df["New Category"])

{'domestic',
 'drug',
 'fraud',
 'kidnap',
 'non_criminal',
 'others',
 'public',
 'sex',
 'suicide',
 'theft'}

In [136]:
df.head()

IncidntNum  Category        Descript DayOfWeek        Date   Time  \
493897         3979  WARRANTS  WARRANT ARREST  Thursday  12/09/2004  16:22   
1209080       10128  WARRANTS  WARRANT ARREST    Sunday  12/18/2005  22:20   
1490233       10736  WARRANTS  WARRANT ARREST  Thursday  07/15/2004  10:18   
1393001       38261  WARRANTS  WARRANT ARREST  Thursday  04/17/2003  22:45   
377806        52205  WARRANTS  WARRANT ARREST  Thursday  02/06/2003  07:20   

        PdDistrict      Resolution                    Address           X  \
493897     BAYVIEW  ARREST, BOOKED     INGALLS ST / HUDSON AV -122.379598   
1209080  INGLESIDE  ARREST, BOOKED  ELLSWORTH ST / ALEMANY BL -122.418969   
1490233   SOUTHERN  ARREST, BOOKED     900 Block of BRYANT ST -122.405927   
1393001   NORTHERN  ARREST, BOOKED        POLK ST / SUTTER ST -122.420120   
377806    SOUTHERN  ARREST, BOOKED     900 Block of BRYANT ST -122.405927   

                 Y                               Location New Category  
493897   37.732467  (37.7324666541275, -122.379598260097)       public  
1209080  37.732208    (37.7322083102079, -122.4189687006)       public  
1490233  37.773427  (37.7734271309539, -122.405926775837)       public  
1393001  37.787757  (37.7877570602182, -122.420120319211)       public  
377806   37.773427  (37.7734271309539, -122.405926775837)       public

In [137]:
# drop redundant columns
df.drop(columns=['Category','Descript','Resolution','Address','X','Y','Location'], inplace=True)
# keep time for now because we will need it to combine it with the sunset/sunrise timings

/Users/clara/anaconda3/envs/data/lib/python3.6/site-packages/pandas/core/frame.py:3930: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


In [138]:
df.head()

IncidntNum DayOfWeek        Date   Time PdDistrict New Category
493897         3979  Thursday  12/09/2004  16:22    BAYVIEW       public
1209080       10128    Sunday  12/18/2005  22:20  INGLESIDE       public
1490233       10736  Thursday  07/15/2004  10:18   SOUTHERN       public
1393001       38261  Thursday  04/17/2003  22:45   NORTHERN       public
377806        52205  Thursday  02/06/2003  07:20   SOUTHERN       public

In [139]:
df.shape

(2057667, 6)

In [140]:
# remove non_criminal
df = df[df["New Category"]!='non_criminal']
df.shape

(1823419, 6)

## 2. Select data between 2012-2017

In [141]:
df['year'] = df['Date'].map(lambda x: pd.to_datetime(x).year)

/Users/clara/anaconda3/envs/data/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [142]:
set(df['year'])

{2003,
 2004,
 2005,
 2006,
 2007,
 2008,
 2009,
 2010,
 2011,
 2012,
 2013,
 2014,
 2015,
 2016,
 2017,
 2018}

In [143]:
df = df[(df['year']>=2012) & (df['year']<=2017)]

In [144]:
set(df['year'])

{2012, 2013, 2014, 2015, 2016, 2017}

In [145]:
# no. of rows of data for 2012-2017
df.shape

(733179, 7)

In [146]:
df.to_csv ('SFCrime_2012-2017.csv', index = None, header=True)

## 3. Count number of unique crimes each day

In [147]:
df.head()

IncidntNum DayOfWeek        Date   Time  PdDistrict New Category  \
75786       1300164  Thursday  02/13/2014  09:00    RICHMOND     domestic   
1185245    10073348  Thursday  11/14/2013  10:20  TENDERLOIN       others   
769622     10358128  Thursday  11/14/2013  10:20  TENDERLOIN       others   
1714235    10733172  Thursday  03/01/2012  13:53     MISSION       public   
2126742    11351210   Tuesday  06/14/2016  08:55     BAYVIEW     domestic   

         year  
75786    2014  
1185245  2013  
769622   2013  
1714235  2012  
2126742  2016

In [148]:
crime_df = df.groupby(['Date', 'DayOfWeek'])['IncidntNum'].agg(['count']).reset_index()
np.sum(crime_df['count'].values)    # sanity check

733179

In [149]:
crime_df.rename(columns={"count": "num_unique_crimes"}, inplace=True)
crime_df.head()

Date  DayOfWeek  num_unique_crimes
0  01/01/2012     Sunday                420
1  01/01/2013    Tuesday                459
2  01/01/2014  Wednesday                385
3  01/01/2015   Thursday                412
4  01/01/2016     Friday                444

## 4. Count number of each crime types

In [150]:
pd.get_dummies(df, columns=['PdDistrict','New Category'])

IncidntNum  DayOfWeek        Date   Time  year  PdDistrict_BAYVIEW  \
75786       1300164   Thursday  02/13/2014  09:00  2014                   0   
1185245    10073348   Thursday  11/14/2013  10:20  2013                   0   
769622     10358128   Thursday  11/14/2013  10:20  2013                   0   
1714235    10733172   Thursday  03/01/2012  13:53  2012                   0   
2126742    11351210    Tuesday  06/14/2016  08:55  2016                   1   
244052     11473002    Tuesday  04/17/2012  10:11  2012                   0   
1354088    12024759     Monday  03/26/2012  05:30  2012                   0   
333040     12036049     Sunday  05/06/2012  02:00  2012                   0   
923224     12603347  Wednesday  03/07/2012  17:00  2012                   0   
248821     13082306     Monday  09/30/2013  13:30  2013                   0   
1507591    13083003     Friday  10/18/2013  08:30  2013                   0   
450453     13088025   Saturday  10/19/2013  22:00  2013                   0   
377600     13103178     Friday  12/06/2013  23:30  2013                   0   
565798     13105685     Friday  12/13/2013  07:30  2013                   0   
1784546    13600061     Friday  12/28/2012  01:00  2012                   0   
1142631    13614363     Friday  10/11/2013  10:30  2013                   0   
358438     13670901     Monday  04/15/2013  16:15  2013                   0   
107052     14056336     Monday  07/07/2014  14:27  2014                   0   
107053     14056336     Monday  07/07/2014  14:27  2014                   0   
60083      15009304     Friday  01/30/2015  19:48  2015                   0   
193238     15046293    Tuesday  05/26/2015  10:45  2015                   1   
193237     15046293    Tuesday  05/26/2015  10:45  2015                   1   
193239     15046293    Tuesday  05/26/2015  10:45  2015                   1   
208231     15059428     Friday  07/10/2015  01:08  2015                   0   
231128     15079313  Wednesday  09/09/2015  22:55  2015                   0   
1825580    15615848     Monday  05/25/2015  21:00  2015                   0   
1829314    15620568    Tuesday  10/20/2015  20:00  2015                   0   
1834153    15626641  Wednesday  09/16/2015  17:32  2015                   0   
1836237    15629231   Thursday  12/03/2015  09:41  2015                   0   
2111417    16003604   Thursday  01/14/2016  15:55  2016                   0   
...             ...        ...         ...    ...   ...                 ...   
963692    970450480     Friday  10/18/2013  15:00  2013                   0   
1018847   970488691     Friday  10/18/2013  15:00  2013                   0   
389771    970715654     Friday  10/18/2013  15:00  2013                   0   
124621    970818101     Monday  08/11/2014  00:01  2014                   0   
1073034   970820015     Friday  10/18/2013  15:00  2013                   0   
630587    970969691     Friday  10/18/2013  15:00  2013                   0   
1443831   971003600     Friday  10/18/2013  15:00  2013                   0   
821128    971339762     Friday  10/18/2013  15:00  2013                   0   
758488    971432429     Friday  10/18/2013  15:00  2013                   0   
719694    971458356     Friday  10/18/2013  15:00  2013                   0   
749614    971661769     Friday  10/18/2013  15:00  2013                   0   
124622    971706672     Monday  08/11/2014  00:01  2014                   0   
1984227   981021761  Wednesday  09/30/2015  23:53  2015                   0   
116408    981356643     Monday  08/18/2014  09:30  2014                   0   
116409    981456168     Monday  08/18/2014  09:30  2014                   0   
116410    981461060     Monday  08/18/2014  09:30  2014                   0   
116411    981469810     Monday  08/18/2014  09:30  2014                   0   
110553    981545614  Wednesday  06/25/2014  00:01  2014                   0   
110554    981579174  Wednesday  06/25/2014  00:01  2014    

## 5. Import housing price dataset

In [151]:
housingprice_df = pd.read_csv('data/HousingPriceFinal.csv')

In [152]:
housingprice_df.head()

Inventory  Median Dom  Median Sale  New Listings Period End  DoW  \
0        7.0        59.5      1500000          15.0    2/29/12  Wed   
1        8.0       145.0       293000           7.0    2/29/12  Wed   
2        4.0        69.0       685000          12.0    2/29/12  Wed   
3       13.0        72.0       316000          19.0    2/29/12  Wed   
4        7.0        59.5       552000          10.0    2/29/12  Wed   

                  Neighborhood PdDistrict  
0  Ashbury Heights / Parnassus       PARK  
1                   Bret Harte    BAYVIEW  
2                  Buena Vista       PARK  
3               Cathedral Hill   NORTHERN  
4                       Cayuga  INGLESIDE

In [153]:
# converting to datetime format, and obtaining columns for 'year', 'month' and 'day'
housingprice_df['year'] = housingprice_df['Period End'].map(lambda x: pd.to_datetime(x).year)
housingprice_df['month'] = housingprice_df['Period End'].map(lambda x: pd.to_datetime(x).month)
housingprice_df['date'] = housingprice_df['Period End'].map(lambda x: pd.to_datetime(x).day)

In [154]:
housingprice_df.head()

Inventory  Median Dom  Median Sale  New Listings Period End  DoW  \
0        7.0        59.5      1500000          15.0    2/29/12  Wed   
1        8.0       145.0       293000           7.0    2/29/12  Wed   
2        4.0        69.0       685000          12.0    2/29/12  Wed   
3       13.0        72.0       316000          19.0    2/29/12  Wed   
4        7.0        59.5       552000          10.0    2/29/12  Wed   

                  Neighborhood PdDistrict  year  month  date  
0  Ashbury Heights / Parnassus       PARK  2012      2    29  
1                   Bret Harte    BAYVIEW  2012      2    29  
2                  Buena Vista       PARK  2012      2    29  
3               Cathedral Hill   NORTHERN  2012      2    29  
4                       Cayuga  INGLESIDE  2012      2    29

In [155]:
housingprice_df=housingprice_df.drop(['Period End','DoW','Neighborhood'], axis=1)

In [156]:
housingprice_df.head()

Inventory  Median Dom  Median Sale  New Listings PdDistrict  year  month  \
0        7.0        59.5      1500000          15.0       PARK  2012      2   
1        8.0       145.0       293000           7.0    BAYVIEW  2012      2   
2        4.0        69.0       685000          12.0       PARK  2012      2   
3       13.0        72.0       316000          19.0   NORTHERN  2012      2   
4        7.0        59.5       552000          10.0  INGLESIDE  2012      2   

   date  
0    29  
1    29  
2    29  
3    29  
4    29

In [157]:
df.head()

IncidntNum DayOfWeek        Date   Time  PdDistrict New Category  \
75786       1300164  Thursday  02/13/2014  09:00    RICHMOND     domestic   
1185245    10073348  Thursday  11/14/2013  10:20  TENDERLOIN       others   
769622     10358128  Thursday  11/14/2013  10:20  TENDERLOIN       others   
1714235    10733172  Thursday  03/01/2012  13:53     MISSION       public   
2126742    11351210   Tuesday  06/14/2016  08:55     BAYVIEW     domestic   

         year  
75786    2014  
1185245  2013  
769622   2013  
1714235  2012  
2126742  2016

In [158]:
df.shape[0]

733179

In [159]:
df['month'] = df['Date'].map(lambda x: pd.to_datetime(x).month)
df['date'] = df['Date'].map(lambda x: pd.to_datetime(x).day)

##### To do:
1) Combine Housing['Period End'] to datetime format
2) Join on these columns: year, month, date, PdDistrict

In [160]:
df.head()

IncidntNum DayOfWeek        Date   Time  PdDistrict New Category  \
75786       1300164  Thursday  02/13/2014  09:00    RICHMOND     domestic   
1185245    10073348  Thursday  11/14/2013  10:20  TENDERLOIN       others   
769622     10358128  Thursday  11/14/2013  10:20  TENDERLOIN       others   
1714235    10733172  Thursday  03/01/2012  13:53     MISSION       public   
2126742    11351210   Tuesday  06/14/2016  08:55     BAYVIEW     domestic   

         year  month  date  
75786    2014      2    13  
1185245  2013     11    14  
769622   2013     11    14  
1714235  2012      3     1  
2126742  2016      6    14

In [161]:
df=df.reset_index(drop=True)
housingprice_df=housingprice_df.reset_index(drop=True)

In [162]:
df.head()

IncidntNum DayOfWeek        Date   Time  PdDistrict New Category  year  \
0     1300164  Thursday  02/13/2014  09:00    RICHMOND     domestic  2014   
1    10073348  Thursday  11/14/2013  10:20  TENDERLOIN       others  2013   
2    10358128  Thursday  11/14/2013  10:20  TENDERLOIN       others  2013   
3    10733172  Thursday  03/01/2012  13:53     MISSION       public  2012   
4    11351210   Tuesday  06/14/2016  08:55     BAYVIEW     domestic  2016   

   month  date  
0      2    13  
1     11    14  
2     11    14  
3      3     1  
4      6    14

In [164]:
df.shape[0]

733179

In [163]:
housingprice_df.head()

Inventory  Median Dom  Median Sale  New Listings PdDistrict  year  month  \
0        7.0        59.5      1500000          15.0       PARK  2012      2   
1        8.0       145.0       293000           7.0    BAYVIEW  2012      2   
2        4.0        69.0       685000          12.0       PARK  2012      2   
3       13.0        72.0       316000          19.0   NORTHERN  2012      2   
4        7.0        59.5       552000          10.0  INGLESIDE  2012      2   

   date  
0    29  
1    29  
2    29  
3    29  
4    29

In [166]:
combinedDF = pd.merge(df, housingprice_df, on=['year','month','date','PdDistrict'],how='left')
combinedDF.sort_values(by=['IncidntNum'])

IncidntNum  DayOfWeek        Date   Time  PdDistrict New Category  \
0          1300164   Thursday  02/13/2014  09:00    RICHMOND     domestic   
1         10073348   Thursday  11/14/2013  10:20  TENDERLOIN       others   
2         10358128   Thursday  11/14/2013  10:20  TENDERLOIN       others   
3         10733172   Thursday  03/01/2012  13:53     MISSION       public   
4         11351210    Tuesday  06/14/2016  08:55     BAYVIEW     domestic   
5         11473002    Tuesday  04/17/2012  10:11    SOUTHERN       public   
6         12024759     Monday  03/26/2012  05:30    NORTHERN     domestic   
7         12036049     Sunday  05/06/2012  02:00    RICHMOND        theft   
8         12603347  Wednesday  03/07/2012  17:00    RICHMOND        theft   
14        13082306     Monday  09/30/2013  13:30    RICHMOND        theft   
13        13082306     Monday  09/30/2013  13:30    RICHMOND        theft   
12        13082306     Monday  09/30/2013  13:30    RICHMOND        theft   
10        13082306     Monday  09/30/2013  13:30    RICHMOND        theft   
9         13082306     Monday  09/30/2013  13:30    RICHMOND        theft   
11        13082306     Monday  09/30/2013  13:30    RICHMOND        theft   
15        13083003     Friday  10/18/2013  08:30     TARAVAL        theft   
16        13088025   Saturday  10/19/2013  22:00    SOUTHERN     domestic   
17        13103178     Friday  12/06/2013  23:30   INGLESIDE        theft   
18        13105685     Friday  12/13/2013  07:30    RICHMOND        fraud   
19        13600061     Friday  12/28/2012  01:00  TENDERLOIN        theft   
20        13614363     Friday  10/11/2013  10:30     CENTRAL        theft   
21        13670901     Monday  04/15/2013  16:15   INGLESIDE        theft   
22        14056336     Monday  07/07/2014  14:27     MISSION        theft   
23        14056336     Monday  07/07/2014  14:27     MISSION       public   
24        15009304     Friday  01/30/2015  19:48        PARK       public   
27        15046293    Tuesday  05/26/2015  10:45     BAYVIEW       kidnap   
25        15046293    Tuesday  05/26/2015  10:45     BAYVIEW          sex   
26        15046293    Tuesday  05/26/2015  10:45     BAYVIEW        theft   
28        15059428     Friday  07/10/2015  01:08   INGLESIDE       others   
29        15079313  Wednesday  09/09/2015  22:55        PARK       others   
...            ...        ...         ...    ...         ...          ...   
869850   971432429     Friday  10/18/2013  15:00  TENDERLOIN       others   
869851   971458356     Friday  10/18/2013  15:00  TENDERLOIN       others   
869852   971661769     Friday  10/18/2013  15:00  TENDERLOIN       others   
869853   971706672     Monday  08/11/2014  00:01  TENDERLOIN       others   
869854   981021761  Wednesday  09/30/2015  23:53    NORTHERN       public   
869855   981021761  Wednesday  09/30/2015  23:53    NORTHERN       public   
869856   981021761  Wednesday  09/30/2015  23:53    NORTHERN       public   
869857   981021761  Wednesday  09/30/2015  23:53    NORTHERN       public   
869858   981021761  Wednesday  09/30/2015  23:53    NORTHERN       public   
869859   981021761  Wednesday  09/30/2015  23:53    NORTHERN       public   
869860   981021761  Wednesday  09/30/2015  23:53    NORTHERN       public   
869861   981021761  Wednesday  09/30/2015  23:53    NORTHERN       public   
869862   981021761  Wednesday  09/30/2015  23:53    NORTHERN       public   
869863   981356643     Monday  08/18/2014  09:30  TENDERLOIN       others   
869864   981456168     Monday  08/18/2014  09:30  TENDERLOIN       others   
869865   981461060     Monday  08/18/2014  09:30  TENDERLOIN       others   
869866   981469810     Monday  08/18/2014  09:30  TENDERLOIN       others   
869867   981545614  Wednesday  06/25/2014  00:01  TENDERLOIN       others   
869868   981579174  Wednesday  06/25/2014  00:01  TENDERLOIN       others   
869869   981630431  Wednesday  06/25/2014  00:01  TENDERLOIN       others   
8698

In [176]:
combinedDF=combinedDF.dropna()

In [177]:
combinedDF

IncidntNum  DayOfWeek        Date   Time PdDistrict New Category  \
9         13082306     Monday  09/30/2013  13:30   RICHMOND        theft   
10        13082306     Monday  09/30/2013  13:30   RICHMOND        theft   
11        13082306     Monday  09/30/2013  13:30   RICHMOND        theft   
12        13082306     Monday  09/30/2013  13:30   RICHMOND        theft   
13        13082306     Monday  09/30/2013  13:30   RICHMOND        theft   
14        13082306     Monday  09/30/2013  13:30   RICHMOND        theft   
60        21549483     Monday  04/30/2012  14:35   SOUTHERN       public   
61        21549483     Monday  04/30/2012  14:35   SOUTHERN       public   
62        21549483     Monday  04/30/2012  14:35   SOUTHERN       public   
63        21549483     Monday  04/30/2012  14:35   SOUTHERN       public   
93       100391111   Saturday  03/31/2012  09:00       PARK     domestic   
94       100391111   Saturday  03/31/2012  09:00       PARK     domestic   
95       100391111   Saturday  03/31/2012  09:00       PARK     domestic   
96       100391111   Saturday  03/31/2012  09:00       PARK     domestic   
97       100391111   Saturday  03/31/2012  09:00       PARK     domestic   
98       100391111   Saturday  03/31/2012  09:00       PARK     domestic   
99       100391111   Saturday  03/31/2012  09:00       PARK     domestic   
100      100391111   Saturday  03/31/2012  09:00       PARK     domestic   
101      100391111   Saturday  03/31/2012  09:00       PARK     domestic   
102      100391111   Saturday  03/31/2012  09:00       PARK     domestic   
103      100391111   Saturday  03/31/2012  09:00       PARK     domestic   
108      100551428  Wednesday  02/29/2012  15:40    BAYVIEW       others   
109      100551428  Wednesday  02/29/2012  15:40    BAYVIEW       others   
110      100551428  Wednesday  02/29/2012  15:40    BAYVIEW       others   
111      100551428  Wednesday  02/29/2012  15:40    BAYVIEW       others   
112      100551428  Wednesday  02/29/2012  15:40    BAYVIEW       others   
113      100551428  Wednesday  02/29/2012  15:40    BAYVIEW       others   
114      100551428  Wednesday  02/29/2012  15:40    BAYVIEW       others   
115      100551428  Wednesday  02/29/2012  15:40    BAYVIEW       others   
258      119009216   Saturday  03/31/2012  02:36    TARAVAL       public   
...            ...        ...         ...    ...        ...          ...   
869637   186013876   Thursday  08/31/2017  20:00   NORTHERN        theft   
869638   186013876   Thursday  08/31/2017  20:00   NORTHERN        theft   
869672   186019153     Sunday  12/31/2017  15:00   NORTHERN        theft   
869674   186019153     Sunday  12/31/2017  15:00   NORTHERN        theft   
869675   186019153     Sunday  12/31/2017  15:00   NORTHERN        theft   
869676   186019153     Sunday  12/31/2017  15:00   NORTHERN        theft   
869677   186019153     Sunday  12/31/2017  15:00   NORTHERN        theft   
869678   186019153     Sunday  12/31/2017  15:00   NORTHERN        theft   
869679   186019153     Sunday  12/31/2017  15:00   NORTHERN        theft   
869680   186019153     Sunday  12/31/2017  15:00   NORTHERN        theft   
869683   186019272     Sunday  12/31/2017  20:00    MISSION        theft   
869684   186019272     Sunday  12/31/2017  20:00    MISSION        theft   
869685   186019272     Sunday  12/31/2017  20:00    MISSION        theft   
869686   186019272     Sunday  12/31/2017  20:00    MISSION        theft   
869687   186019272     Sunday  12/31/2017  20:00    MISSION        theft   
869700   186024293     Sunday  12/31/2017  23:00    CENTRAL        theft   
869701   186024293     Sunday  12/31/2017  23:00    CENTRAL        theft   
869702   186024293     Sunday  12/31/2017  23:00    CENTRAL        theft   
869703   186024293     Sunday  12/31/2017  23:00    CENTRAL        theft   
869704   186024293     Sunday  12/31/2017  23:00    CENTRAL        theft   
869705   186024293     Sunday  12/31/2017  23:00    CENT

In [179]:
combinedDF.to_csv('data/CrimeAndHousing.csv', index = None, header=True)

## 6. Import Eviction Dataset

In [223]:
eviction_df = pd.read_csv('data/eviction_minus.csv',header=0)

In [224]:
eviction_df.shape

(40091, 30)

In [225]:
pd.set_option('display.max_columns', 30)
eviction_df.head()

Eviction.ID                        Address           City State  \
0     M190330   1800 Block Of Market  Street  San Francisco    CA   
1     M183362  1100 Block Of Mission  Street  San Francisco    CA   
2     M183157       0 Block Of Brady  Street  San Francisco    CA   
3     M182998    0 Block Of Stillman  Street  San Francisco    CA   
4     M182997    0 Block Of Stillman  Street  San Francisco    CA   

   Eviction.Notice.Source.Zipcode File.Date  Non.Payment  Breach  Nuisance  \
0                         94102.0   2/15/19        False   False      True   
1                         94103.0  12/20/18        False   False      True   
2                         94103.0  11/21/18        False    True     False   
3                         94107.0  10/29/18        False   False      True   
4                         94107.0  10/29/18        False    True     False   

   Illegal.Use  Failure.to.Sign.Renewal  Access.Denial  Unapproved.Subtenant  \
0        False                    False          False                 False   
1        False                    False          False                 False   
2        False                    False          False                 False   
3        False                    False          False                 False   
4        False                    False          False                 False   

   Owner.Move.In  Demolition  Capital.Improvement  Substantial.Rehab  \
0          False       False                False              False   
1          False       False                False              False   
2          False       False                False              False   
3          False       False                False              False   
4          False       False                False              False   

   Ellis.Act.WithDrawal  Condo.Conversion  Roommate.Same.Unit  Other.Cause  \
0                 False             False               False        False   
1                 False             False               False        False   
2                 False             False               False        False   
3                 False             False               False        False   
4                 False             False               False        False   

   Late.Payments  Lead.Remediation  Development  Good.Samaritan.Ends  \
0          False             False        False                False   
1          False             False        False                False   
2          False             False        False                False   
3          False             False        False                False   
4          False             False        False                False   

  Constraints.Date  Supervisor.District Neighborhoods...Analysis.Boundaries  \
0              NaN                  8.0                        Hayes Valley   
1              NaN                  6.0                     South of Market   
2              NaN                  6.0                             Mission   
3              NaN                  6.0      Financial District/South Beach   
4              NaN                  6.0      Financial District/South Beach   

                                    Location PdDistrict  
0    (37.7712923782185, -122.42433694872517)   SOUTHERN  
1   (37.77779030910167, -122.41283433258292)   SOUTHERN  
2  (37.772548429191886, -122.42008435391327)   SOUTHERN  
3    (37.78258788740649, -122.3952163164336)   SOUTHERN  
4    (37.78258788740649, -122.3952163164336)   SOUTHERN

In [226]:
eviction_df=eviction_df.drop(['Address','City','State','Location'], axis=1)

In [227]:
eviction_df.head()

Eviction.ID  Eviction.Notice.Source.Zipcode File.Date  Non.Payment  Breach  \
0     M190330                         94102.0   2/15/19        False   False   
1     M183362                         94103.0  12/20/18        False   False   
2     M183157                         94103.0  11/21/18        False    True   
3     M182998                         94107.0  10/29/18        False   False   
4     M182997                         94107.0  10/29/18        False    True   

   Nuisance  Illegal.Use  Failure.to.Sign.Renewal  Access.Denial  \
0      True        False                    False          False   
1      True        False                    False          False   
2     False        False                    False          False   
3      True        False                    False          False   
4     False        False                    False          False   

   Unapproved.Subtenant  Owner.Move.In  Demolition  Capital.Improvement  \
0                 False          False       False                False   
1                 False          False       False                False   
2                 False          False       False                False   
3                 False          False       False                False   
4                 False          False       False                False   

   Substantial.Rehab  Ellis.Act.WithDrawal  Condo.Conversion  \
0              False                 False             False   
1              False                 False             False   
2              False                 False             False   
3              False                 False             False   
4              False                 False             False   

   Roommate.Same.Unit  Other.Cause  Late.Payments  Lead.Remediation  \
0               False        False          False             False   
1               False        False          False             False   
2               False        False          False             False   
3               False        False          False             False   
4               False        False          False             False   

   Development  Good.Samaritan.Ends Constraints.Date  Supervisor.District  \
0        False                False              NaN                  8.0   
1        False                False              NaN                  6.0   
2        False                False              NaN                  6.0   
3        False                False              NaN                  6.0   
4        False                False              NaN                  6.0   

  Neighborhoods...Analysis.Boundaries PdDistrict  
0                        Hayes Valley   SOUTHERN  
1                     South of Market   SOUTHERN  
2                             Mission   SOUTHERN  
3      Financial District/South Beach   SOUTHERN  
4      Financial District/South Beach   SOUTHERN

In [228]:
eviction_df['year']=eviction_df['File.Date'].map(lambda x: pd.to_datetime(x).year)
eviction_df['month'] = eviction_df['File.Date'].map(lambda x: pd.to_datetime(x).month)
eviction_df['date'] = eviction_df['File.Date'].map(lambda x: pd.to_datetime(x).day)


In [222]:
eviction_df.head()

Eviction.ID  Eviction.Notice.Source.Zipcode File.Date  Non.Payment  Breach  \
0     M190330                         94102.0   2/15/19        False   False   
1     M183362                         94103.0  12/20/18        False   False   
2     M183157                         94103.0  11/21/18        False    True   
3     M182998                         94107.0  10/29/18        False   False   
4     M182997                         94107.0  10/29/18        False    True   

   Nuisance  Illegal.Use  Failure.to.Sign.Renewal  Access.Denial  \
0      True        False                    False          False   
1      True        False                    False          False   
2     False        False                    False          False   
3      True        False                    False          False   
4     False        False                    False          False   

   Unapproved.Subtenant  Owner.Move.In  Demolition  Capital.Improvement  \
0                 False          False       False                False   
1                 False          False       False                False   
2                 False          False       False                False   
3                 False          False       False                False   
4                 False          False       False                False   

   Substantial.Rehab  Ellis.Act.WithDrawal  Condo.Conversion  \
0              False                 False             False   
1              False                 False             False   
2              False                 False             False   
3              False                 False             False   
4              False                 False             False   

   Roommate.Same.Unit  Other.Cause  Late.Payments  Lead.Remediation  \
0               False        False          False             False   
1               False        False          False             False   
2               False        False          False             False   
3               False        False          False             False   
4               False        False          False             False   

   Development  Good.Samaritan.Ends Constraints.Date  Supervisor.District  \
0        False                False              NaN                  8.0   
1        False                False              NaN                  6.0   
2        False                False              NaN                  6.0   
3        False                False              NaN                  6.0   
4        False                False              NaN                  6.0   

  Neighborhoods...Analysis.Boundaries PdDistrict  year  month  date  
0                        Hayes Valley        NaN  2019      2    15  
1                     South of Market        NaN  2018     12    20  
2                             Mission        NaN  2018     11    21  
3      Financial District/South Beach        NaN  2018     10    29  
4      Financial District/South Beach        NaN  2018     10    29